### PR TITLE
Use template to handle image import

### LINKED ISSUE #8388: [IDEA] Support image drag & drop (cut & paste) for markdown tiddlers
**Is your feature request related to a problem? Please describe.**
Yes. The editor didn't support dragging & dropping (cutting & pasting) images when the type field is `text/markdown`. Adding `[all[current]type[text/markdown]]` to `$:/config/Editor/EnableImportFilter` didn't work too. 

(I found that the message did appear for markdown tiddlers after modifying `$:/config/Editor/EnableImportFilter`, but the import dialog didn't appear. The import dialog only appears when the content type is `text/vnd.tiddlywiki` or empty.)

图片

**Describe the solution you'd like*
Markdown tiddlers support dragging and dropping (cutting and pasting) images like tiddlywiki tiddlers.

### PR DESCRIPTION
**Pull request by Leilei332**
This PR supports pasting image in markdown in a way which is different from #8463 and #8486 .

It creates the tiddler `$:/core/ui/FileImportTemplate` to be used as template. The UX tiddler transcludes the template tiddler with four fields set:

* `image` : image link template
* `file` : normal link template
* `regexp-search` and `regexp-escape` : regular expression to escape characters

Closes #8388 .

@saqimtiaz this PR explores a different way to reduce code duplication.

### CODE REVIEW

File: core/ui/FileImportTemplate.tid [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+	</$let>
+</$list>
+<<closePopupActions>>
+\end
+
+\procedure buttons()
+<$button class="tc-btn-invisible tc-small-gap-right" actions=<<closePopupActions>> ><<lingo Listing/Cancel/Caption>></$button>
+<$button class="tc-btn-invisible" message="tm-perform-import" param=<<importTitle>> actions=<<postImportActions>> ><<lingo Listing/Import/Caption>></$button>
+\end
+
```
Comment (pmario) [inline core/ui/FileImportTemplate.tid:30]: please indent the **following** code

**Comment by Jermolene** [comment]
Thanks @Leilei332 I'm catching up after a break. It would be very helpful if we could figure out the best path forward with @saqimtiaz.

**Comment by Jermolene** [comment]
Thanks @Leilei332. Are we now in a position to decide between this PR and #8463 and #8486?

**Comment by Leilei332** [comment]
#8463 is ready to be closed. For #8486 I wonder whether it is ready to be converted from draft and merged.

**Comment by saqimtiaz** [comment]
@Leilei332 thank you for taking a crack at this. Note that with this implementation, while drag and drop support is enabled for markdown tiddlers, the user experience for editing markdown tiddlers will be broken for users until and unless they have the markdown plugin installed. The ability to edit markdown tiddlers in TiddlyWiki without having the markdown plugin to render them is a valid though uncommon use case that we need to be able to cater to.

 The advantage of the approach in #8486 is that the editing experience for markdown remains robust and the drag and drop support is present even without the need for having the markdown plugin installed.

**Comment by Leilei332** [comment]
IMO we had better move `$:/language/Docs/Types/text/markdown` out of the markdown plugin if we moves the file import handler for markdown out of the markdown plugin.

**Comment by Jermolene** [comment]
Thanks @Leilei332 – should this be closed now that #8486 has been merged?

**Comment by Leilei332** [comment]
Actually I prefer this one, since #8486 mixed the handler for wikitext links and markdown links in one tiddler. IMO they should exist in two seperate tiddlers, which is clearer.

**Comment by Leilei332** [comment]
I think this should be a pluggable mechamism like editor toolbars. I do think there may be plugins like AsciiDoc plugin and reStructuredText plugin in the future.

IMO the mechamism should be customizable just like tiddlywiki's other user interface.

**Comment by saqimtiaz** [comment]
I would fully support adding a pluggable mechanism that depends on the mime type of the tiddler should the need arise, ideally one that does not require a separate editor toolbar component tiddler for each type.

 My recommendation would be to open an issue for the idea and revisit it if/when the need arises, otherwise we introduce a fair bit of complexity potentially for no benefit.

**Comment by Jermolene** [comment]
Hi @Leilei332 I am not familiar with AsciiDoc and reStructuredText, do they use image formats that could plausibly fit into such a pluggable framework?

**Comment by Leilei332** [comment]
The syntax for reStructuredText is:

```rst
.. image:: gnu.png
   :alt: Alternative text
```

For the AsciiDoc one:

```asciidoc
image::sunset.jpg[Sunset]
```

Since they are all lightweight markup languages like markdown, it won't be difficult to fit them in tiddlywiki. IMO making the mechanism pluggable should the direction for the development of the image import mechanism.

**Comment by saqimtiaz** [comment]
My hesitation to add support for a pluggable mechanism at this time comes from the lack of real world examples and demands to inform the design.

 If we are going to pursue this at this time, we should also consider making sure the mechanism is flexible enough to support greater customizability within each mime type. Such as being able to support other user configurable syntaxes for wikitext. For example, in a tiddler tagged 'gallery' use the custom procedure 'gallery-image' and in other tiddlers use the procedure 'imagebox'.

 I am not sure that the mechanism proposed in this PR is configurable and flexible enough for such use cases, and would welcome further exploration in this area.

**Comment by Leilei332** [comment]
I think we can close this one and discuss it in #8662.

**Pull request closed without merging**